src/mpi/init/initthread.c

```diff
@@ -76,120 +76,143 @@ int MPI_Init_thread(int *argc, char ***argv, int required, int *provided)
+     * done now.  Try to keep this as small as possible.  Only things
+     * that *REALLY* cannot be done before device initialization
+     * should come here. */
+    /**********************************************************************/
+
+    /* pairtypes might need device hooks to be activated so the device
+     * can keep track of their creation.  that's why we need to do
+     * this after the device initialization.  */
+    mpi_errno = MPIR_Datatype_commit_pairtypes();
     MPIR_ERR_CHECK(mpi_errno);
 
+    /* FIXME: this is for rlog, and should be removed when we delete rlog */
+    MPII_Timer_init(MPIR_Process.comm_world->rank, MPIR_Process.comm_world->local_size);
+
     MPII_post_init_memory_tracing();
     MPII_init_dbg_logging();
-    /* if --enable-debuginfo is configured, prepare for debug info data.
-     * if MPIEXEC_DEBUG is set in environment, wait for debugger until MPIR_debug_gate is set to 1.
-     */
     MPII_Wait_for_debugger();
 
-    /* dup comm_self and creates progress thread (if needed) */
-    mpi_errno = MPII_init_async();
+
+    /**********************************************************************/
+    /* Section 6: simply lets the device know that we are done with
+     * the full initialization, in case it wants to do some final
+     * setup. */
+    /**********************************************************************/
+
+    mpi_errno = MPID_InitCompleted();
     MPIR_ERR_CHECK(mpi_errno);
 
-    /* create fine-grained mutexes */
-    MPIR_Thread_CS_Init();
+    MPL_atomic_store_int(&MPIR_Process.mpich_state, MPICH_MPI_STATE__POST_INIT);
 
-    /* ---- MPID_Post_init --------------------------------------------- */
-    /* connect to remote processes is has parent */
-    if (MPIR_Process.has_parent) {
-        mpi_errno = MPID_Init_spawn();
-    }
 
-    /* Let the device know that the rest of the init process is completed */
-    mpi_errno = MPID_InitCompleted();
+    /**********************************************************************/
+    /* Section 7: we are finally ready to start the asynchronous
+     * thread, if needed. */
+    /**********************************************************************/
 
-    /* ---- Exit CS --------------------------------------------------- */
-    MPII_init_thread_and_exit_cs();
+    /* Reset isThreaded to the actual thread level */
+    MPIR_ThreadInfo.isThreaded = (MPIR_ThreadInfo.thread_provided == MPI_THREAD_MULTIPLE);
 
-    MPL_atomic_store_int(&MPIR_Process.mpich_state, MPICH_MPI_STATE__POST_INIT);
+    mpi_errno = MPII_init_async();
+    MPIR_ERR_CHECK(mpi_errno);
 
     if (provided)
         *provided = MPIR_ThreadInfo.thread_provided;
     return mpi_errno;
 
   fn_fail:
     /* --BEGIN ERROR HANDLING-- */
-    /* signal to error handling routines that core services are unavailable */
-    MPL_atomic_store_int(&MPIR_Process.mpich_state, MPICH_MPI_STATE__PRE_INIT);
-    MPII_init_thread_failed_exit_cs();
+    MPII_thread_mutex_destroy();
```

Comment:
If we just want to abort here, maybe we don't need worry about destroying mutexes?

Reply:
That's what the `MPIR_Err_return_comm` would do when there's no error handler setup.  So we are already doing that.

Reply:
There is talk in the MPI-Forum that we want allow the option of non-default error handler for MPI_Init, so I guess we *cannot* assume it is always abort.

But my question is do we need destroy mutexes here? If we remove that, then `init_mutex` become truly independent components.

Reply:
I don't think we should support such error returning before the MPI state is initialized.  It's a nightmare to maintain.